lld/ELF/InputSection.cpp

```diff
@@ -898,10 +898,16 @@ void InputSection::relocateNonAlloc(uint8_t *buf, ArrayRef<RelTy> rels) {
   const TargetInfo &target = *elf::target;
   const auto emachine = config->emachine;
   const bool isDebug = isDebugSection(*this);
-  const bool isDebugLocOrRanges =
-      isDebug && (name == ".debug_loc" || name == ".debug_ranges");
   const bool isDebugLine = isDebug && name == ".debug_line";
-  std::optional<uint64_t> tombstone;
+  std::optional<uint64_t> tombstone, debugTombstone;
```

Comment:
Seems unfortunate that the `tombstone` and `debugTombstone` handling is distinct here - rather than `debugTombstone` being a default behavior that could be expressed via the `tombstone`/`deadRelocInNonAlloc` functionality. (eg: for the 32/64 bit issue - I guess there's no way for a user to, on the command line, express "use max int/all-1s in this section, for both 32 bit and 64 bit (or other size) relocations"?)

(or, alternatively, perhaps it'd be simpler to remove the generic-but-not-quite-sufficient-to-express-these-situations `deadRelocInNonAlloc` flag and have something narrower that describes the specific tombstoning strategies we want to support, pre-canned (v5-and-below compatible and v6-and-above/standard tombstoning, basically))

Reply:
`type == target.symbolicRel` is a proxy for whether the relocation is 64-bit on ELF64 (`symbolicRel`).
Previously, the entries in debug sections that need tombstone values are all `symbolicRel` since they reference a text section (and for generalization, we don't use the small code model assumption).
`.debug_names` is a new special case where its `.long .Ltu_begin0 # Type unit 0` reference is 32-bit.

A natural extension of the existing code looks like this.

`-z dead-reloc-in-nonalloc=` was initially designed to support both 32-bit and 64-bit absolute relocations.

---

Actually, x86-64 is unique in discerning signed/unsigned 32-bit absolute relocations (R_X86_64_32/R_X86_64_32S).
We can special case it instead:
```
if (emachine == EM_X86_64 && type == R_X86_64_32)
  value = static_cast<uint32_t>(value);
```

(AArch64/PPC64/RISC-V/etc don't need this.)

I have made the simplication.

Reply:
I was mostly hoping there was some way to write the code more generally - I was inspired by @cmtice's first prototype where she had a switch over many different relocation types to handle tombstone values in different sizes of relocations (which aren't used currently, but I can appreciate the desire to have generic/general code in the linker) - so I was wondering if we could have that sort of generalized support without having to special case all the different sizes of relocations.

But, anyway, that's just my 2c. But sounds like lld's architecture doesn't make that feasible at this time - so be it.

Reply:
Yes, the arch-neutral property has pros and cons.

I benchmarked this function with a similar one in mold. IIRC mold's is 2.x as fast as this lld implementation.

[template <class ELFT, class RelTy> InputSection::relocateNonAlloc](https://github.com/llvm/llvm-project/blob/5b7a7ec5a2106772de90a59c52e9fac7481f7e8a/lld/ELF/InputSection.cpp#L888) has performance 
problems:

* arch-specific checks in generic code: `R_386_GOTPC` workaround, x86 `R_SIZE`, RISC-V ADD/SUB relocations, (future) RISC-V ULEB128
* virtual function overhead (`X86, X86_64, ARM, AArch64`, etc : `TargetInfo`)
  + `addend += target.getImplicitAddend(bufLoc, type);`
  + `RelExpr expr = target.getRelExpr(type, sym, bufLoc);`
  + `target.relocateNoSym(bufLoc, type, SignExtend64<bits>(sym.getVA(addend)));`

I wanted to eliminate the overhead so I made [InputSection::relocateAlloc](https://github.com/llvm/llvm-project/commit/685b21255315e699aa839d93fe71b37d806c90c2) arch-specific but it's difficult to change `relocateNonAlloc` as `relocateNonAlloc` is a template (cannot be virtual).